model/PlantFeatExtNew.py

```diff
@@ -0,0 +1,104 @@
+        """
+        super().__init__()
+        num_channels = [num_hidden] * (num_levels - 1) + [embedding_size]
+        self.tcn = TemporalConvNet(2048, num_channels, kernel_size, dropout)
+
+    def forward(self, x: torch.Tensor):
+        """
+        :param x: input tensor of size (NxTx2048),
+                where N is the batch size, T is the sequence length and 2048 is the input embedding dim
+        :return: tensor of size (N x T x embedding_size) where out[:,t,:] is the output given all values up to time t
+        """
+        # transpose each sequence so that we get the correct size for the TemporalConvNet
+
+        # print(x.shape) --> torch.Size([4, 57, 1, 2048])
+        #
+        x = torch.stack([m.t() for m in x])
+
+        # Forward through TCN:
+        out = self.tcn(x)
+
+        # undo the previous transpose
+        return torch.stack([m.t() for m in out])
+
+
+class PlantFeatureExtractor(nn.Module):
+    def __init__(self,
+                 embedding_size: int = parameters.net_features_dim,
+                 **mods: Dict[str, Union[int, float]]):
+        super().__init__()
+
+        # all modalities
+        self.mods = list(mods.keys())
+
+        # image features were already extracted by inception - read features with DataLoader
+
+        # a dictionary for the feature extractors for each modality
+        self.mod_extractors: Dict[str, nn.Module] = dict()
+
+        # create a feature extractor with the inputted params for each param modality - TCN
+        for mod in mods.keys():
+            self.mod_extractors[mod] = TCNextractor(**mods[mod])
+            self.add_module(f'TCN_{mod}_feat_extractor', self.mod_extractors[mod])
+
+        # final feature extractor - linear FC
+        d_in = parameters.tcn_embedding_size * len(self.mods)
+        # Exp0: tcn_embedding_size(=128)* 5 mods = 640//
+        # Exp2: tcn_embedding_size(=128)*6 mods = 768
+        d_out = embedding_size  # 512
+        # h = int((d_in + d_out) / 2)  # 640
+
+        # print('D_in  = ', d_in, 'D_out  = ', d_out, ' H = ', h)
+        self.final_feat_extractor = nn.Linear(d_in, d_out)  # FC Linear last layer
+
+    def forward(self, **x: torch.Tensor):
+        """
+        :param x: input of type mod=x_mod for each modality where each x_mod is of shape
+                    (NxT_modx2048),
+                    where the batch size N is the same over all mods and others are mod-specific
+        :return: a feature vector of shape (N x embedding_size)
+        """
+
+        # extract the features for each mod using the corresponding feature extractor
+        mod_feats = {}
+        for mod in self.mods:
+            # print(mod, 'x[mod].shape ', x[mod].shape) # torch.Size([4, 57, 1, 2048])
+            x[mod] = torch.squeeze(x[mod])
+            # print('x[mod].shape after torch.squeeze(x[mod]) ', x[mod].shape)  #torch.Size([4, 57, 2048])
+
+            # FORWARD:
+            mod_feats[mod] = self.mod_extractors[mod](x[mod])
```

Comment:
- [ ] use comprehension
```suggestion
        mod_feats = {mod: self.mod_extractors[mod](torch.squeeze(x[mod])) for mod in self.mods}
```